### PR TITLE
Provcau 84 custom display components

### PR DESCRIPTION
**Pull request by bri-gaia**

### CODE REVIEW

File: app/lib/Parsers/ExpressionParser/ExpressionVisitor.php [hunk 1]
```diff
@@ -72,6 +72,15 @@ protected function initializeFunctions() {
 			return implode($vs_glue, $va_args);
 		};
 
+		$custom = function($method, ...$args){
+			if (is_callable($method)){
```
Comment (kehh) [inline app/lib/Parsers/ExpressionParser/ExpressionVisitor.php]: We should probably have an allow list of these in config (app.conf) somewhere cos this is a pretty big security hole thinking about it.

File: app/models/ca_bundle_displays.php [hunk 2]
```diff
@@ -2003,6 +2005,54 @@ public function getPlacementsInDisplay($options=null) {
 		return $placements_in_display;
 	}
 	# ------------------------------------------------------
+	private function getCustomDisplayComponents(&$va_available_bundles, $t_instance, $vs_table, $vb_show_tooltips, $vs_format) {
+		// Search for Custom Display Components
+		$va_additional_settings = [];
+		$va_file_list = [];
+		$vs_custom_displays_path = __DIR__ . "/CustomDisplayComponents/";
```
Comment (kehh) [inline app/models/ca_bundle_displays.php]: Can you turn this into a custom hook call?
Something similar to hookRegisterTaskQueuePluginDirectories and then basically just trawl through the the returned list of directories. Then we can put the actual components in a plugin.

File: app/models/ca_bundle_displays.php [hunk 3]
```diff
@@ -2277,6 +2327,17 @@ public function getDisplayValue($po_result, $pn_placement_id, $options=null) {
...
+
+		if (preg_match('/^custom_display_component/', $va_bundle_bits[1])) {
+			$vs_component = str_replace('custom_display_component_', '', $va_bundle_bits[1]);
+			$vs_filename = __CA_APP_DIR__ . "/models/CustomDisplayComponents/{$va_bundle_bits[0]}/{$vs_component}.php";
+			if (file_exists($vs_filename)) {
+				$vs_template = file_get_contents($vs_filename);
+			} else {
+				$vs_template = 'ERROR - missing template ' . $vs_filename;
+			}
+			error_log($vs_filename);
```
Comment (kehh) [inline app/models/ca_bundle_displays.php]: move log into else statement and include something like 'Failed to find missing template'

File: app/models/ca_bundle_displays.php (hunk 2 above)
Comment (kehh) [reply app/models/ca_bundle_displays.php]: So a plugin from PROVCAU-85 can implement a method like:

```php
<?
    public function hookRegisterCustomDisplayComponentDirectories($pa_directories = []):array
    {
        $pa_directories[] = __DIR__ . '/CustomDisplayComponents/';
        return $pa_directories;
    }
```

**Review by kehh — CHANGES_REQUESTED** [review]

File: app/models/ca_bundle_displays.php (hunk 3 above)
Comment (bri-gaia) [reply app/models/ca_bundle_displays.php]: this is actually just debugging that I missed >.<

File: app/lib/Parsers/ExpressionParser/ExpressionVisitor.php (hunk 1 above)
Comment (bri-gaia) [reply app/lib/Parsers/ExpressionParser/ExpressionVisitor.php]: Removed this change from this PR so we can address these issues individually as they are evolving past the original implementations.

See https://github.com/gaiaresources/providence/pull/7

**Comment by bri-gaia** [comment]
I can't shake the feeling something is off here.

Should we be adding hooks to allow plugins to define display components and then implementing custom templates in our plugin? seems more messy, but also more inline with the current functionality. Is this a completely separate idea that could be shelved?

**Comment by kehh** [comment]
> I can't shake the feeling something is off here.
> 
> Should we be adding hooks to allow plugins to define display components and then implementing custom templates in our plugin? seems more messy, but also more inline with the current functionality. Is this a completely separate idea that could be shelved?

~~I think it's neat for the plugin(s) to do everything - certify things as safe and implement those things.~~

**Comment by bri-gaia** [comment]
As per discussion with Kehan, going with this approach.

**Pull request merged**